A43-sdk-authorization.md

```diff
@@ -0,0 +1,721 @@
+  		  "monitoring and error message generation.",
+  		"type": "string"
+  	},
+  	"deny_rules": {
+  		"description": "List of deny rules to match. If a request matches any of the"
+  		  "deny rules, then it will be denied. If none of the deny rules matches or"
+  		  "there are no deny rules, the allow rules will be evaluated.",
+  		"type": "array",
+  		"items": {
+  			"$ref": "#/definitions/rule"
+  		}
+  	},
+  	"allow_rules": {
+  		"description": "List of allow rules to match. The allow rules will only be"
+  		  "evaluated after the deny rules. If a request matches any of the allow"
+  		  "rules, then it will allowed. If none of the allow rules matches, it will"
+  		  "be denied.",
+  		"type": "array",
+  		"items": {
+  			"$ref": "#/definitions/rule"
+  		}
+  	}
+  },
+  "required": ["name", "allow_rules"]
+}
+```
+
+#### Details
+
+gRPC SDK authorization policy has a list of *allow rules* and a list of *deny rules*.
+The following sequence is followed to make an authorization decision -
+1. Check for a match in *deny rules*, if a match is found, the request will be denied.
+   If no match is found in deny rules, or there are no deny rules execute next step.
+2. Check for a match in *allow rules*, if a match is found, the request will be allowed.
+   Note that allow rules is a required field here, if not present, policy is invalid.
+3. If no match is found in *allow rules*, we deny the request.
+
+Each *rule* has the following semantics -
+1. Each *rule* has a *name*, a *source* and a *request* to match against. For a *rule*
+   to match, both *source* and *request* must match.
+   - If both *source* and *request* are empty, the *rule* always matches. It is a
+     wildcard that can be used in a deny list to deny all RPCs or in an allow list to
+     accept all RPCs except those matching in the deny list.
+   - If only *source* is empty, we evaluate against the *request* fields and apply
+     that to any user (See example below). Similarly, if only *request* is empty, we
+     evaluate against the *source* fields and apply that to any action.
+2. Each *source* could contain a list of *principals*. The *principals* are ORed
+   together, i.e. it matches if one of them matches.
+   Sequence of steps to evaluate each principal from config -
+   1. If TLS is not used, matching fails.
+   2. If there is no client certificate, we get a match if principal is an empty
+      string.
+   3. If we have a client certificate, we check against certificate contents. A
+      match is found if principal matches URI SANs from the certificate. If
+      there are no URI SANs in certificate, or no match was found, we check
+      against DNS SANs. Similarly, if certificate has no DNS SANs or match wasn't
+      found, we check against Subject field from certificate.
+   
+   Consider the case where the *principals* list is empty, then we only check for
+   step 1 above, i.e. rule matches for any authenticated user.
```

Comment:
If only step 1 is checked, then that means all users accessing the service over TLS are permitted. For only authenticated users, it looks like you should use `*`.

Reply:
Done.